### PR TITLE
[stable/postgesql] Added custom init container

### PR DESCRIPTION
**Pull request by kel00s**
For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

#### What this PR does / why we need it:
This PR allow to use additional/custom init container for additional tasks which can be done before pod is running. One example is to restore DB from backup where is hard to estimate time and setup correct initial delay for health checks. 

#### Special notes for your reviewer:
@cedricgc 

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] title of the PR contains starts with chart name e.g. `[stable/chart]`

### CODE REVIEW

File: stable/postgresql/templates/_helpers.tpl [hunk 1]
```diff
@@ -74,6 +74,30 @@ Also, we can't use a single if because lazy evaluation is not an option
...
+        {{- printf "%s/%s:%s" .Values.global.imageRegistry $repositoryName $tag -}}
+    {{- else -}}
+        {{- printf "%s/%s:%s" $registryName $repositoryName $tag -}}
+    {{- end -}}
+{{- else -}}
+    {{- printf "%s/%s:%s" $registryName $repositoryName $tag -}}
+{{- end -}}
+{{- end -}}
+
+
```
Comment (carrodher) [inline stable/postgresql/templates/_helpers.tpl]: It seems here there is an extra line

File: stable/postgresql/values.yaml [hunk 2]
```diff
@@ -57,6 +57,23 @@ volumePermissions:
   securityContext:
     runAsUser: 0
 
+##
+## Init container can be separately added to Master and Slave stateful set by changing respectively
+## initContainer.masterEnabled and initContainer.slavesEnabled values
+##
+
```
Comment (carrodher) [inline stable/postgresql/values.yaml]: Just to be consistent with the rest of the comments, this line is not needed, you can add the variables just after the `##`

File: stable/postgresql/values.yaml [hunk 3]
```diff
@@ -57,6 +57,23 @@ volumePermissions:
   securityContext:
     runAsUser: 0
 
+##
```
Comment (carrodher) [inline stable/postgresql/values.yaml]: This is not needed, at least we are not using it in other comments

File: stable/postgresql/templates/statefulset.yaml [hunk 4]
```diff
@@ -79,6 +79,86 @@ spec:
...
+          {{- range .Values.initContainer.command }}
+          - {{ . }}
+          {{- end }}
+        {{- end}}
+        securityContext:
+          runAsUser: {{ .Values.volumePermissions.securityContext.runAsUser }}
+        volumeMounts:
+        - name: data
+          mountPath: {{ .Values.persistence.mountPath }}
+        env:
```
Comment (carrodher) [inline stable/postgresql/templates/statefulset.yaml:99]: Why do you need all the env. variables to be passed to the master initcontainer but not to the slave one?

File: stable/postgresql/README.md [hunk 5]
```diff
@@ -156,6 +156,15 @@ The following tables lists the configurable parameters of the PostgreSQL chart a
...
+| `initContainer.image.registry`                | PostgreSQL init container Image registry                                                                               | `docker.io`                                                 |
+| `initContainer.image.repository`              | PostgreSQL init container Image name                                                                                   | `busybox`                                                   |
+| `initContainer.image.tag`                     | PostgreSQL init container Image tag                                                                                    | `latest`                                                    |
+| `initContainer.image.pullPolicy`              | PostgreSQL init container Image pull policy                                                                            | `Always`                                                    |
+| `initContainer.image.command`                 | Command to start init container                                                                                        | `- /bin/bash`                                               |
+
+
```
Comment (carrodher) [inline stable/postgresql/README.md]: Unnecessary blank line

**Review by carrodher — COMMENTED** [review]
Thanks for the PR!
How do you see adding a section in the README including information about how to use it? (Apart from the list of variables that you already included)

File: stable/postgresql/templates/statefulset.yaml (hunk 4 above)
Comment (kel00s) [reply stable/postgresql/templates/statefulset.yaml:99]: Yeah, I guess forgot to add those to slaves.

**Comment by kel00s** [comment]
> Thanks for the PR!
> How do you see adding a section in the README including information about how to use it? (Apart from the list of variables that you already included)

Yeah, sure. I've made quick note.

File: stable/postgresql/templates/_helpers.tpl (hunk 1 above)
Comment (kel00s) [reply stable/postgresql/templates/_helpers.tpl]: Removed.

File: stable/postgresql/values.yaml (hunk 2 above)
Comment (kel00s) [reply stable/postgresql/values.yaml]: Removed.

File: stable/postgresql/values.yaml (hunk 3 above)
Comment (kel00s) [reply stable/postgresql/values.yaml]: Removed.

File: stable/postgresql/README.md (hunk 5 above)
Comment (kel00s) [reply stable/postgresql/README.md]: Removed.

**Comment by juan131** [comment]
Hi @kel00s 

Thanks so much for the contribution!

I don't fully understand why it's necessary to add this functionality to the chart. We already provide a way to add custom **init scripts** that allows you to perform a series of operations on PostgreSQL. Please take a look to the parameters:

- initdbScripts
- initdbScriptsConfigMap
- initdbScriptsSecret

More information can be found at https://github.com/helm/charts/tree/master/stable/postgresql#initialize-a-fresh-instance

IMHO, adding custom init containers adds too much complexity to the chart (which is already very complex)

**Pull request closed without merging**